lib/tutf8e:  refactoring for flb_encoding (WIP)

bluebike opened the pull request:
**WIP**:  in process of enable encoding first few changes to tutf8e encoding library.

----

**Summary**
<!-- Provide summary of changes -->

* separate _encoding_* functions removed.
* length_functions return TUTF8E_SAME if no changes needed in string
* multiple modes modes for handling invalid chars:  
  * KEEP: (default) just encode byte to utf8 
  * IGNORE: skip character 
  * FAIL: fail if invalid chars
  * REPLACEMENT:  replace invalid character with unicode 0xFFFD
  * QUESTION:  relace invalid characters with '?' char.
* tests functions made for invalid char handling. 

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [ ] Example configuration file for the change
- [ ] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [ ] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

bluebike commented:
@nigels-com  would like to see some comments of my changes... 
Mainly I tried to make this handle invalid chars + get info is string doesn't need changes (optimization).

nigels-com commented:
I'll take a closer look.

Seems like a significant rework, some further discussion of the goals and rationale would be helpful, I think.

nigels-com commented:
My first question from a `fluent-bit` perspective is if there is a realistic need for all the modes?
The expectation is to configure this per input?  Which mode is the intended default?
From an embedded perspective, I'd like to only link the code/mode that is needed for a particular application.

bluebike commented:
@nigels-com  Probably overkill having different modes.
But different users of fluent-bit could have different requirements. 
Having different modes doesn't really add that much code  I think, mostly test code.

nigels-com commented:
@bluebike You're saying that the advantage of REPLACEMENT or QUESTION modes are that they'll flag message streams that are not actually encoded as expected, but at least the ASCII parts will succeed.  Do you feel that KEEP and IGNORE are also important modes?

nigels-com commented:
I'm not convinced the TUTF8E_SAME mode is worth the additional complexity.  Have you been able to measure any actual throughput, memory or CPU improvements?

nigels-com commented:
As an aside, I filed some issues over on the tutf8e repo, as reminders to myself.

https://github.com/nigels-com/tutf8e/issues/8
https://github.com/nigels-com/tutf8e/issues/6

nigels-com commented on lib/tutf8e/src/tutf8e.c:
```diff
@@ -3,121 +3,200 @@
-/* return TUTF8E_OK for success                                */
+/* return >= 0 size of encoded string                          */
 
-int tutf8e_buffer_encode(const uint16_t *table, const char *input, size_t ilen, char *output, size_t *olen)
```
One comment about performance assumptions.

The formulation here with _olen_ it to avoid exceeding the output buffer without pre-computing the length of the encoded output.  Without checking, this change _assumes_ and _requires_ the output buffer to be sufficiently large. 

If we assume there is "ample" space in the output buffer, it should not be necessary to do two steps.  But we do need to check for when we do (occasionally) run out of space.

bluebike replied:
yes.. your original code has has check for output size... that is better. 
But having this dual purpose **\*olen** is al little bit problematic: input=size of buffer, output=size of output. 
Ok could go with that.

nigels-com commented on lib/tutf8e/src/tutf8e.c:
```diff
@@ -3,121 +3,200 @@
 {
   int ret;
-  size_t ilen = 0;
+  size_t ilen = strlen(input);
```
Now we're doing _three_ passes of the input, rather than two!  I hope the input is not too long...

bluebike replied:
hum.. that was left from some changes... to be revmoed.

nigels-com commented on lib/tutf8e/src/tutf8e.c:
```diff
@@ -3,121 +3,200 @@
-/* return TUTF8E_INVALID if input character is not convertable                       */
-/* return TUTF8E_OK for success                                                      */
+/* return TUTF8E_INVALID if input character is not convertable  TUTF8E_FLAG_INV_FAIL */
+/* return TUTF8E_SAME    if string doesnt' change  (but copied) */
```
Suggest deleting the words "but copied" here.   tutf8e_string_length does not copy.

bluebike replied:
yeah... (as this is still WIP) comments needs to looking around.

bluebike commented:
Ok.. .did some changes...   
* **\*olen** used as giving output buffer length  + returning  length of output utf8 string. 
* some comments fixed. 
* removed extra strlen
* added test for too short output buffer returning TUTF8E_TOOLONG

nigels-com commented:
@bluebike Please take a look at PR #2326 to see if this arrangement strikes a good balance of features and "tinyness".

Rather than the various iconv modes, it adds support for a UTF-8 proxy for each invalid input character.
If the proxy is NULL, the API will continue erroring for invalid input.

bluebike commented:
(this goes nowhere)
